### PR TITLE
An exact match of the original Acey Ducey game using some new features in java after 1.8

### PR DESCRIPTION
**Pull request by jnellis**
This is as close as I can get to a near exact duplicate, character for character, space for space, of the original game as interpreted by the Vintage Basic interpreter, including error messages.  Try running them side by side and spot any differences.

There are many notes throughout the code pointing to some comments at the bottom that explain either new features of java or the original behavior of the basic program.  

FYI, you can run .java source code files directly without having to compile them first now. That was introduced in JDK11.  This version of AceyDucey does require JDK17 features to run.

### CODE REVIEW

**Comment by coding-horror** [comment]
Fantastic! If this a simpler layout for the Java versions we should use this harness in all of them! 🙌

**Comment by jnellis** [comment]
As I look at some of the other Java submissions they are broken up into multiple files which I think is a missed opportunity as java now allows to run straight .java files like script files, without pre-compiling.  Java even allows a shebang as the first line so they can be treated as .sh files. But in order for these to run like a script there must not be dependencies on other files, such as an actual generic harness that one would `extend`, because it would force the need to pre-compile.  

There should probably be an area for guidance; to maybe show snippets like how to handle text input without throwing a stacktrace at the user.

**Comment by journich** [comment]
I think in some cases it might be a good idea to put the java code into one file.  For example the games that are only using a second file to invoke the game.  Putting the main method into the game code file would make good sense.

I'd say for other games that use extra classes for other reasons, its probably not good programming practice try and cram these into one file purely to be able to run without compilation. 

To me the project is aimed at programmers first and foremost, and less so end users.  So being able to run without compilation is useful, but not essential.  Programmers will be able to download and install a JDK.  IDE's make running code pretty trivial as well.

For anything complicated, instructions can be provided. 

Teaching good programming practices is essential here, in my opinion.

**Comment by jnellis** [comment]
@journich  People are going to do what they want. Most java learners in college or where ever, are taught one class, one file, and so it's like the law for them now.  My point is that (most of) these were once one page scripts and that they are simple enough that they should remain that if possible.  It's hard enough to not exceed the line count of the original BASIC program.  There's room for more than one submission per program per language if the solutions are different enough.  You could be more functional or way more OO and those would both be good learning examples of the spectrum that you can program in java. We still have yet to see the Enterprise AceyDucey solution. :)

Also, side note, your version of AceyDucey is kinda borked. If you choose to replay you are not given another 100, your account remains at zero. There are a couple other out of order display issues.

**Comment by journich** [comment]
Or sure there is absolutely room for more than one submission per language if there is a valid reason to do it.  Jeff and myself would welcome that.

I've targeted JDK 8/JDK 11 with my submissions since thats what is mostly used by companies, but some of the newer features in later versions are pretty cool.  Now that JDK 17 is the LTS version perhaps thats the default I should be using.

I just wouldn't like to see code being crammed into a single file just for the sake of it.  Even if its found in production code sometimes.

Would love to see functional and OO versions for all games, that would be cool and very useful as a learning tool.

Thanks for the heads up on AceyDucey - I'll check it out and fix the issue.  I believe it was one of, if not my first submission and I probably didn't test enough.

Out of interest what program are you using to test the basic program?  I was using Chipmunk Basic but have just found and started using Vintage Basic on my mac.

**Comment by jnellis** [comment]
>I just wouldn't like to see code being crammed into a single file just for the sake of it. Even if its found in production code sometimes.

I agree. I think we are "on the same page here".  And by production code, I'll just mention that lowly ol java.util.HashMap has 14 inner classes at 2500loc. In the grand scope of things, regardless of what language is used, there is a line count of the original BASIC code. Should there be some kind of proportionality to that metric? If the original BASIC weighs in at 100loc and a submission is 1000loc regardless of how many files it spans, that might be an eyebrow raiser.  I found it very difficult to stay under the original loc if any kind of readability was a concern but I ended up not much over (minus notes), about 50% the original. 

The original BASIC AceyDucey is horrible!! There is a lot of room to refactor it down to something simpler and thus reduce the line count even further. (maybe that should a target language.) GOTO really was powerful. I remember entering in these programs in the 80's, there was always the "did I make a typo or is it a bug" scenario going on. After combing over and deciding its not a typo, you would then have to wait for future issues for the errata of bug fixes, if they were even deemed important enough to displace advertising space!

I'm using the Vintage Basic as recommended. I tried a number of online interpreters at first because I was lazy but none of them worked.  Most of my BASIC days were 40 years ago on the Trash80, 2e, and c64 as a kid.  This was a fun to revisit.

**Comment by coding-horror** [comment]
> Would love to see functional and OO versions for all games, that would be cool and very useful as a learning tool.

I wouldn't go too overboard on the OO here.. we don't want the "Enterprise Edition" of these games, just nice clean functional _modern programming_ versions ;)

**Comment by jnellis** [comment]
| we don't want the "Enterprise Edition" of these games

You mean like this?
https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition

**Comment by coding-horror** [comment]
yes that's exactly what I meant lol

**Comment by journich** [comment]
> > Would love to see functional and OO versions for all games, that would be cool and very useful as a learning tool.
> 
> I wouldn't go too overboard on the OO here.. we don't want the "Enterprise Edition" of these games, just nice clean functional _modern programming_ versions ;)

I'm not talking about enterprise code when I refer to OO, just standard Java. Many of the smaller games need very little in the way of OO, but some of the bigger ones can and do utilise OO principles in the Java and C# and probably other versions to make the game easier to understand than the original code.

**Comment by journich** [comment]
> Should there be some kind of proportionality to that metric? If the original BASIC weighs in at 100loc and a submission is > 1000loc regardless of how many files it spans, that might be an eyebrow raiser. I found it very difficult to stay under the > original loc if any kind of readability was a concern but I ended up not much over (minus notes), about 50% the original.

I have not attempted to match loc in any way. I've just written the code the way I normally would, without introducing unnecessarily complex code in the Java version.

I think readability is more important that reducing the number of lines of code, especially given how unreadable most of these basic games are!

> 
> The original BASIC AceyDucey is horrible!! There is a lot of room to refactor it down to something simpler and thus reduce the line count even further. (maybe that should a target language.) GOTO really was powerful.

Many, perhaps most are pretty bad.  I've ported about 30 games to Java so far and most of the basic games are pretty hard to follow, or at least very time consuming to make sure you have thoroughly tested the game.

Maybe the VB versions of the code can focus on lines of code?  I've used C# quite a bit, but not VB, so it would actually be a fun exercise to refactor these games in VB and make them readable.

I might swing over and pick up VB and see what thats like these days.

It took me quite a while to get out of using GOTO when I eventually left basic for other languages!

> I remember entering in these programs in the 80's, there was always the "did I make a typo or is it a bug" scenario going >on. After combing over and deciding its not a typo, you would then have to wait for future issues for the errata of bug >fixes, if they were even deemed important enough to displace advertising space!

Same here.  In my case I lived in a small country town and bought the books because the magazines were not available at the time in Australia, so I never got the benefit of future bug fixes in an errata.

Still it was one of the most enjoyable periods of my life, typing the code in and getting things to work (or not), enhancing things, and so on.

> 
> I'm using the Vintage Basic as recommended. I tried a number of online interpreters at first because I was lazy but none of them worked. Most of my BASIC days were 40 years ago on the Trash80, 2e, and c64 as a kid. This was a fun to revisit.

So far I'm liking Vintage basic.   My background was Microbee Z80 16K (Australian company) then Amstrad CPC464, then C64, and swinging over to PC.
